### PR TITLE
Guard artifact uploading in wandb with ddp barriers

### LINKED ISSUE #90: WandB error due to multiple artifacts with the same ID
When running a baseline resnet50 model on imagenet, I encountered this error:
```
wandb: ERROR Error while calling W&B API: Error 1062: Duplicate entry '6394579-1' for key 'unique_artifact_collection_membership_version' (<Response [409]>)
Exception in thread Thread-7:
Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.8/dist-packages/wandb/filesync/step_upload.py", line 50, in _thread_body
    self._handle_event(event)
  File "/usr/local/lib/python3.8/dist-packages/wandb/filesync/step_upload.py", line 79, in _handle_event
    self._maybe_commit_artifact(job.artifact_id)
  File "/usr/local/lib/python3.8/dist-packages/wandb/filesync/step_upload.py", line 161, in _maybe_commit_artifact
    self._api.commit_artifact(artifact_id)
  File "/usr/local/lib/python3.8/dist-packages/wandb/sdk/internal/internal_api.py", line 2235, in commit_artifact
    response = self.gql(mutation, variable_values={"artifactID": artifact_id})
  File "/usr/local/lib/python3.8/dist-packages/wandb/sdk/lib/retry.py", line 102, in __call__
    result = self._call_fn(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/wandb/sdk/internal/internal_api.py", line 147, in execute
    six.reraise(*sys.exc_info())
  File "/usr/local/lib/python3.8/dist-packages/six.py", line 719, in reraise
    raise value
  File "/usr/local/lib/python3.8/dist-packages/wandb/sdk/internal/internal_api.py", line 141, in execute
    return self.client.execute(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/wandb/vendor/gql-0.2.0/gql/client.py", line 52, in execute
    result = self._get_result(document, *args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/wandb/vendor/gql-0.2.0/gql/client.py", line 60, in _get_result
    return self.transport.execute(document, *args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/wandb/vendor/gql-0.2.0/gql/transport/requests.py", line 39, in execute
    request.raise_for_status()
  File "/usr/local/lib/python3.8/dist-packages/requests/models.py", line 953, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 409 Client Error: Conflict for url: https://api.wandb.ai/graphql
```

I've asked the WandB folks and they think it's from an attempted upload of an artifact with the same ID as another. The recent addition of artifact uploading from `run_directory` seems to be causing this, so PR #89 will disable it by default, but we need to verify that artifact uploads are working as expected.

### PR DESCRIPTION
**Pull request by ravi-mosaicml**
Ensuring that the run directory is not being modified while wandb is uploading artifacts.

Hopefully closes #90

### CODE REVIEW

File: tests/test_logger.py [hunk 1]
```diff
@@ -90,3 +91,15 @@ def get_mock_tqdm(position: int, *args, **kwargs):
...
         assert mock_tqdm.update.call_count == trainer._eval_subset_num_batches
         mock_tqdm.close.assert_called_once()
+
+
+@pytest.mark.parametrize("world_size", [
+    pytest.param(1),
+    pytest.param(2, marks=pytest.mark.world_size(2)),
+])
+@pytest.mark.timeout(10)
+def test_wandb_logger(mosaic_trainer_hparams: TrainerHparams, world_size: int):
```
Comment (jbloxham) [inline tests/test_logger.py:101]: Is this just to verify that the logger doesn't crash?

**Review by jbloxham — APPROVED** [review]
Looks good to me! I'm curious whether this is a class of errors that might crop up later for other callbacks. Let's keep this issue in mind.

File: tests/test_logger.py (hunk 1 above)
Comment (ravi-mosaicml) [reply tests/test_logger.py]: Yea, we didn't have a wandb logger test before. This just verifies that it's initialized properly, etc...

**Comment by ravi-mosaicml** [comment]
This issue _should_ only appear on processes that need the run directory to be locked -- which at this point is just wandb and the run directory uploader.

**Pull request merged**